### PR TITLE
switch to `:retry` 2-arity fun, add support for {:delay, milliseconds}

### PR DESCRIPTION
**Pull request by DaTrader**
As noted here https://github.com/wojtekmach/req/pull/203#issuecomment-1693289313, I made the agreed changes except for the one changing the behavior of the function when it returns `true`. That part I left as it was i.e. it will have delay fetched from the `retry-after` header only if the status code is 429. Again, my advice is not to force users to double check their existing code base when switching to the 2-arity function.

### CODE REVIEW

File: lib/req/response.ex [hunk 1]
```diff
@@ -141,11 +141,11 @@ defmodule Req.Response do
       when is_binary(key) and is_binary(value) do
     %{response | headers: List.keystore(response.headers, key, 0, {key, value})}
   end
-
+
```
Comment (wojtekmach) [inline lib/req/response.ex:144]:
```suggestion

```

File: lib/req/steps.ex [hunk 2]
```diff
@@ -1405,14 +1414,35 @@ defmodule Req.Steps do
...
                   "got: #{inspect(other)}"
       end
 
-    if retry? do
-      retry(request, response_or_exception)
-    else
-      {request, response_or_exception}
+    case retry do
+      {:delay, delay} ->
+        unless Req.Request.get_option(request, :retry_delay) do
```
Comment (wojtekmach) [inline lib/req/steps.ex:1423]: please change to if/else

File: lib/req/steps.ex [hunk 3]
```diff
@@ -1405,14 +1414,35 @@ defmodule Req.Steps do
...
+      true ->
+        retry(request, response_or_exception)
+
+      retry when retry in [false, nil] ->
+        {request, response_or_exception}
+    end
+  end
+
+  defp apply_retry(fun, request, response_or_exception) do
+    case Function.info(fun, :arity) do
```
Comment (wojtekmach) [inline lib/req/steps.ex:1439]: I believe Function.info is slow, please use:

```elixir
fun when is_function(fun, 1) ->
  # ...

fun when is_function(fun, 2) ->
  # ...
```

File: lib/req/steps.ex (hunk 2 above)
Comment (DaTrader) [reply lib/req/steps.ex:1423]: but why?

**Comment by wojtekmach** [comment]
Sorry about the conflicts, I tried resolving some but not sure if I didn’t make things worse.

**Comment by DaTrader** [comment]
> Sorry about the conflicts, I tried resolving some but not sure if I didn’t make things worse.

You removed my tests and on my computer the tests don't pass any more.

**Comment by DaTrader** [comment]
It's funny that your CI is passing.

**Comment by wojtekmach** [comment]
If the production code is roughly what you wanted I can take it from here but of course if you’d like to keep going that’s appreciated too. I think I did a merge with GitHub UI so your changes should be recoverable. Sorry about that.

**Comment by DaTrader** [comment]
I'm not following you. What should we do now? Should I bring my tests back or? It's not a problem, just tell me what you expect me to do.

**Comment by wojtekmach** [comment]
I didn’t mean to remove anything it was a botched merge. Sorry for interfering. Please proceed as you see fit and I’ll review it in a couple days.

**Comment by DaTrader** [comment]
> Please proceed as you see fit and I’ll review it in a couple days.

No reason to wait. I just commented a single test that you added at the last moment and it all works now. Please  merge and then you can work on that test later.

**Comment by wojtekmach** [comment]
Thank you!

**Comment by DaTrader** [comment]
Thank you!

**Pull request merged**